### PR TITLE
[BEAM-5983] Create Combine load test

### PR DESCRIPTION
**Pull request by lgajowy**
@aaltay could you take a look at this too?

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/) [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | --- | ---

### CODE REVIEW

File: sdks/java/testing/load-tests/src/main/java/org/apache/beam/sdk/loadtests/CombineLoadTest.java [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+    PCollection<KV<byte[], byte[]>> input =
+        pipeline
+            .apply("Read input", SyntheticBoundedIO.readFrom(sourceOptions))
+            .apply("Collect metrics", ParDo.of(new MetricsMonitor(METRICS_NAMESPACE)));
+
+    for (int i = 0; i < options.getFanout(); i++) {
+      applyStepIfPresent(input, format("Step: %d", i), syntheticStep)
+          .apply(
+              format("Convert to BigInteger: %d", i), MapElements.via(new ByteValueToBigInteger()))
+          .apply(format("Combine: %d", i), combiner);
```
Comment (lgajowy) [inline sdks/java/testing/load-tests/src/main/java/org/apache/beam/sdk/loadtests/CombineLoadTest.java]: I need some feedback about what Combine modes should we test. Are Top.perKeyLargest and Mean.perKey combine operations enough?

Comment (aaltay) [reply sdks/java/testing/load-tests/src/main/java/org/apache/beam/sdk/loadtests/CombineLoadTest.java]: Let's add sum and count.

File: sdks/java/testing/load-tests/src/main/java/org/apache/beam/sdk/loadtests/CombineLoadTest.java [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+  private static final String METRICS_NAMESPACE = "combine";
+
+  private enum CombinerType {
+    TOP_LARGEST,
+    MEAN
+  }
+
+  /** Pipeline options specific for this test. */
+  interface Options extends LoadTestOptions {
```
Comment (aaltay) [inline sdks/java/testing/load-tests/src/main/java/org/apache/beam/sdk/loadtests/CombineLoadTest.java]: Could we have some standard configs to run this load test? Perhaps a wrapper to run this with a set of parameters

I can think of:
Uniform
Normal
Onekey
Fanout 256 (could be uniform)
(+ maybe a few hot keys in a larger distribution.)

File: sdks/java/testing/load-tests/src/main/java/org/apache/beam/sdk/loadtests/CombineLoadTest.java (hunk 1 above)
Comment (lgajowy) [reply sdks/java/testing/load-tests/src/main/java/org/apache/beam/sdk/loadtests/CombineLoadTest.java]: ok

File: sdks/java/testing/load-tests/src/main/java/org/apache/beam/sdk/loadtests/CombineLoadTest.java (hunk 2 above)
Comment (lgajowy) [reply sdks/java/testing/load-tests/src/main/java/org/apache/beam/sdk/loadtests/CombineLoadTest.java]: I agree. This also applies to other tests (Gbk, CoGbk, ParDo etc). It's probably worth to add default parameter sets in a separate PR (I find it a cross-cutting concern). Wdyt?

Comment (aaltay) [reply sdks/java/testing/load-tests/src/main/java/org/apache/beam/sdk/loadtests/CombineLoadTest.java]: Sounds good.

Comment (lgajowy) [reply sdks/java/testing/load-tests/src/main/java/org/apache/beam/sdk/loadtests/CombineLoadTest.java]: Created JIRA for this: https://issues.apache.org/jira/browse/BEAM-6065

**Comment by lgajowy** [comment]
@aaltay applied your suggestions. PTAL again.

**Comment by lgajowy** [comment]
Run Java PreCommit

**Comment by lgajowy** [comment]
Thanks!

**Pull request merged**